### PR TITLE
pack more fields in datatype

### PR DESCRIPTION
**Pull request by vtjnash**
As Jeff noted recently, more fields can be moved to reduce the data size (I think removing `.names` had the biggest effect on sysimg size, but all seemed helpful). I've also packed some Bool fields too.

### CODE REVIEW

**Comment by KristofferC** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

Just curious about the "damage".

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/41018#issuecomment-851235673) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/50fc0b9_vs_b46df09/report.md). cc @maleadt

**Comment by vtjnash** [comment]
Looks like a couple "key" packages (SymbolServer, AxisArrays, JuliaInterpreter), but not a major impact.

File: base/compiler/typeutils.jl [hunk 1]
```diff
@@ -56,6 +58,38 @@ function isknownlength(t::DataType)
     return isdefined(va, :N) && va.N isa Int
 end
 
+# Compute the minimum number of initialized fields for a particular datatype
+# (therefore also a lower bound on the number of fields)
+function datatype_min_ninitialized(t::DataType)
+    isabstracttype(t) && return 0
+    if t.name === NamedTuple_typename
+        names, types = t.parameters[1], t.parameters[2]
```
Comment (JeffBezanson) [inline base/compiler/typeutils.jl:66]: We could just call `fieldcount` in this case.

Comment (JeffBezanson) [reply base/compiler/typeutils.jl:66]: Hm, I guess not, since we want 0 instead of an error for abstract types...

File: src/julia.h [hunk 2]
```diff
@@ -1025,14 +1034,7 @@ JL_DLLEXPORT jl_svec_t *jl_compute_fieldtypes(jl_datatype_t *st JL_PROPAGATES_RO
...
 {
-    jl_svec_t *names = st->names;
-    if (!names)
-        names = st->name->names;
-    return names;
-}
-STATIC_INLINE jl_sym_t *jl_field_name(jl_datatype_t *st, size_t i) JL_NOTSAFEPOINT
-{
-    return (jl_sym_t*)jl_svecref(jl_field_names(st), i);
+    return st->name->names;
```
Comment (JeffBezanson) [inline src/julia.h:1037]: Maybe assert here that it's not a named tuple?

**Review by JeffBezanson — APPROVED** [review]

File: base/compiler/typeutils.jl (hunk 1 above)
Comment (vtjnash) [reply base/compiler/typeutils.jl:66]: yes, that is why it is reimplemented here

File: src/julia.h (hunk 2 above)
Comment (vtjnash) [reply src/julia.h:1037]: I figured it would assert due to the null pointer anyways, so it was unnecessary to enforce it

**Pull request merged**